numpyro/examples/vae.py

```diff
@@ -0,0 +1,152 @@
+
+
+@jit
+def binarize(rng, batch):
+    rng, = random.split(rng, 1)
+    return rng, random.bernoulli(rng, batch).astype(batch.dtype)
+
+
+def main(args):
+    encoder_init, encode = encoder(args.hidden_dim, args.z_dim)
+    decoder_init, decode = decoder(args.hidden_dim, 28 * 28)
+    opt_init, opt_update = optimizers.adam(args.learning_rate)
+    svi_init, svi_update, svi_eval = svi(model, guide, elbo, opt_init, opt_update,
+                                         encode=encode, decode=decode, z_dim=args.z_dim)
+    svi_update = jit(svi_update)
+    rng = PRNGKey(0)
+    train_init, train_fetch = load_dataset(MNIST, batch_size=args.batch_size, split='train')
+    test_init, test_fetch = load_dataset(MNIST, batch_size=args.batch_size, split='test')
+    num_train, train_idx = train_init()
+    _, encoder_params = encoder_init((args.batch_size, 28 * 28))
+    _, decoder_params = decoder_init((args.batch_size, args.z_dim))
+    params = {'encoder': encoder_params, 'decoder': decoder_params}
+    rng, sample_batch = binarize(rng, train_fetch(0, train_idx)[0])
+    opt_state = svi_init(rng, (sample_batch,), (sample_batch,), params)
+    rng, = random.split(rng, 1)
+
+    @jit
+    def epoch_train(opt_state, rng):
+        def body_fn(i, val):
+            loss_sum, opt_state, rng = val
+            rng, batch = binarize(rng, train_fetch(i, train_idx)[0])
+            loss, opt_state, rng = svi_update(i, opt_state, rng, (batch,), (batch,),)
+            loss_sum += loss
+            return loss_sum, opt_state, rng
+
+        return lax.fori_loop(0, num_train, body_fn, (0., opt_state, rng))
+
+    @jit
+    def eval_test(opt_state, rng):
+        def body_fun(i, val):
+            loss_sum, rng = val
+            rng, = random.split(rng, 1)
+            rng, batch = binarize(rng, test_fetch(i, test_idx)[0])
+            loss = svi_eval(opt_state, rng, (batch,), (batch,)) / len(batch)
+            loss_sum += loss
+            return loss_sum, rng
+
+        loss, _ = lax.fori_loop(0, num_test, body_fun, (0., rng))
+        loss = loss / num_test
+        return loss
+
+    def reconstruct_img(epoch):
+        img = test_fetch(0, test_idx)[0][0]
+        plt.imsave(os.path.join(RESULTS_DIR, 'original_epoch={}.png'.format(epoch)), img, cmap='gray')
+        _, test_sample = binarize(rng, img)
+        params = optimizers.get_params(opt_state)
+        z_mean, z_var = encode(params['encoder'], test_sample.reshape([1, -1]))
+        z = dist.norm(z_mean, z_var).rvs(random_state=rng)
+        img_loc = decode(params['decoder'], z).reshape([28, 28])
+        plt.imsave(os.path.join(RESULTS_DIR, 'recons_epoch={}.png'.format(epoch)), img_loc, cmap='gray')
+
+    for i in range(args.num_epochs):
+        t_start = time.time()
+        num_train, train_idx = train_init()
+        _, opt_state, rng = epoch_train(opt_state, rng)
+        rng, = random.split(rng, 1)
+        num_test, test_idx = test_init()
+        test_loss = eval_test(opt_state, rng)
```

Comment:
nit: I think it is better to split rng into 2 (1 to train and update rng, 1 for eval) to avoid a confusion that: rng here is used for eval and the train of next epoch.

Reply:
Done.